### DISCUSSION TITLE
1. 네트워크의 기초

### DISCUSSION THREAD

**Discussion by jindream6128** [24tngus/CS_STUDY#28] [Network]
`노드` - `노드` 들이 링크로 연결되어 있는 집합. 

노드 : 서버, 라우터, 스위치 등 네트워크 장치

링크(엣지) : 유무선 연결 매체(wifi, lan)

![ㅂ](https://github.com/24tngus/CS_STUDY/assets/101094583/fdee27a5-b0f1-4d60-8b8a-b6413829bcc8)

세상은 모든 노드들과 링크되어 있다.

`트래픽`: 노드사이의 링크 내에 흐르는 데이터의 양!

트래픽 | 처리량 차이 

- 트래픽이 많아졌다 → 흐르는 데이터가 많아졌다.
- 처리량이 많아졌다 → 처리되는 데이터가 많아졌다.

`처리량:` 링크 내에서 성공적으로 전달된 데이터 양 → 보통 얼만큼의 트래픽을 처리하는지 나타냄

많은 트래픽을 처리한다 → 많은 처리량을 가진다 

처리량은 → 트래픽의 양, 네트워크 장치간의 대역폭, 중간에서 발생하는 에러, HW스펙에 영향을 받게된다.

`대역폭` : BandWidth → 주어진 시간 동안 네트워크 연결을 통해 흐를 수 있는 최대 비트 수 (최대로 처리할수 있는 트래픽 = 최대의 처리량)

`RTT(Round Trip Time)`왕복 지연 시간  : 두 장치 사이에 데이터가 송 수신되는데 걸리는 시간 → `ping`

## 네트워크 토폴로지 : 버스 스타 트리 링 메시

→ 노드와 링크들이 구성된 모양

## 1. 버스

![ㅈ](https://github.com/24tngus/CS_STUDY/assets/101094583/8b9c1f21-d767-43a0-8fe9-091b24bd1baf)

`특징` : 1개의 회선, 여러노드 → 추가, 삭제가 쉬워서 비용이 적다

`장점` : 소규모 네트워크, 하나의 노드에 문제가 생겨도 이상이 없다.

`단점` : 메인 링크가 망가지면 큰 문제, 트래픽이 정체 될 수 있다. 

## 2. 스타

![ㄷ](https://github.com/24tngus/CS_STUDY/assets/101094583/41d857c0-a7d2-4c57-a8bc-f071dff23142)

`특징` : 중앙에 노드에 모두 연결되어있어 → 추가 삭제가 쉬움

`장점` : 중앙 외의 노드에는 문제가 생겨도 상관 없다, 안정성이 높다. 

`단점` : 중앙 노드에 집중되어 있다.

## 3. 트리

![ㄱ](https://github.com/24tngus/CS_STUDY/assets/101094583/c2d6a156-e02c-4365-956c-9683fd2188e5)

`특징` : 트리형태로서 계층적 토폴로지라고도 한다. 버스 + 스타 형으로서 중앙집중되어있는 스타형의 중심 노드를 제거 시 까다롭다.

`장점` :  리프노드의 확장에 용이, 리프노드의 에러는 영향이 없다

`단점` :  특정한 노드에 트래픽이 집중되면 하위노드에 영향을 끼치고, 루트노드에 문제가 발생하면 큰영향

## 4. 링

![ㅅ](https://github.com/24tngus/CS_STUDY/assets/101094583/94832f97-723d-463f-a6a7-d2b4322248a1)

`특징` : 추가,삭제 쉬움

`장점` : 노드가 많아져도 데이터의 손실이 없다, 권한에 따라 특정 노드에 데이터를 전달하지 않을 수 있다 → 토큰이용.

`단점` : 링크 또는 노드가 하나만 에러가 생겨도 문제가 발생한다. 토큰이 없다면 통신을 할수 없다. 

## 5. 메시

![ㅛ](https://github.com/24tngus/CS_STUDY/assets/101094583/2a3e5bca-7b56-4c04-8a8b-bf296f6bcb2d)

`특징` : 그물망 형태로서, 노드 추가 삭제가 어렵다 

풀 메시 토폴로지의 경우 n(n-2)/2 의 회선이 필요하다

`장점` : 안정성이 높고 트래픽을 분산할 수 있다.

`단점` : 회선이 비효율적으로 많아, 구축비용이 높다.

### 병목현상

트래픽에 의해 데이터의 흐름이 제한되는 상황!

→ 특정 상황에 따라 토폴로지를 맞추어 구성

`유니캐스트` 

1:1 통신으로서, 대표적으로는 HTTP 통신이 있다.

→ 가장 일반적
![ㅕ](https://github.com/24tngus/CS_STUDY/assets/101094583/2b76a2f9-ccae-4ecc-b5c0-cf41cd7aa7d7)

`멀티 캐스트`

1:N 통신으로서, 연결된 모든 노드들에게 데이터를 전달하는 것이 아닌 특정 그룹에만 전달.

![ㅑ](https://github.com/24tngus/CS_STUDY/assets/101094583/e13c6ec5-2142-41cb-b29c-3944f2298f51)

`브로드캐스트`

1:N 통신으로서, 그룹이 아닌 연결되어 있는 모든 노드에게 데이터를 전달.

![ㅐ](https://github.com/24tngus/CS_STUDY/assets/101094583/641abd86-d301-4506-843a-efa8fd744db0)

## 네트워크의 분류 : LAN MAN WAN

LAN → MAN → WAN

분류 기준 : 속도나 크기

`LAN(local area network)` : 근거리 통신망으로서, 안정성과 높은 속도를 가진다.

하나의 논리적인 주소인 IP 를 기반으로 여러개의 물리적인 주소MAC 주소로 구별하는 네트워

![ㅁ](https://github.com/24tngus/CS_STUDY/assets/101094583/709f6b2d-d8ab-4fe6-8e55-6bb085f7f9db)

`MAN(metropolitan area network)` : 대도시 통신망으로서 2개 이상의 LAN으로 구성, 라우터나 브릿지 등으로 연결

![ㄴ](https://github.com/24tngus/CS_STUDY/assets/101094583/4d08e991-fd14-4ff5-b24f-f3a7c4ea5101)

`WAN(wide area network)` : 광역 통신망으로서 국가와 국가의 통신망을 뜻하며, 인터넷 이라고도 한다. 수많은 라우터를 거쳐 나라가 연결되어 있음.

![ㅇ](https://github.com/24tngus/CS_STUDY/assets/101094583/3e2a51bc-a77b-43cc-8321-6f4a1d0e251a)

Q) 회사 외부 -> 내부로 네트워크가 들어는 경우
1. `외부 부터의 요청이수신`되면, 외부 네트워크와 내부 네트워크에 물려있는 장비에 도달하게됩니다! 이때 물려잇는 장비들의 경우 방화벽, 라우터, 로드 밸런스 등이 있습니다.
2. `방화벽 검사` : 외부 요청을 검사하여 보안 정책에 따라 허용되는 요청인지 검사하고, 로그들을 남기게 됩니다.
3. `주소 변환` (NAT) : 방화벽이 외부 요청을 허용하면 네트워크 주소 변환 (NAT : Network Address Translation)이 수행되고, 외부에서 사용한 IP 주소와 포트 번호가 내부에서 사용하는 사설 IP 주소와 포트 번호로 변환되게 됩니다.
4. `라우팅`: 변환된 내부 주소와 포트 번호를 기반으로 라우터가 내부 네트워크로 요청을 전달합니다. 이때 라우터는 패킷의 목적지 IP주소를 확인하여 어떤 내부 네트워크 장비로 전송될지 결정 합니다.
5. `내부 서버 또는 서비스 접근 `: 외부의 요청이 라우팅에 의해 내부로 전달되면 해당 요청에 따라 요청이 처리됩니다. 예를 들어 웹 서비스의 경우 HTTP요청이라면 웹 서버에서 해당 요청을 처리 합니다.
6. `응답 전송`: 내부 서버에서 요청을 처리한 뒤 다시 외부 클라이언트에게 전달되게 됩니다.

**Comment by mingkyeongg** [comment]
내 주변 네트워크, 기업들은 어떤 토폴로지를 사용할까 궁금한네요!

**Comment by sujin-park0607** [comment]
음 회사마다 네트워크 토폴로지는 누가 결정하고 설계하는지도 궁금합니다! 인프라 담당자들이 처음에 설계하는건가?

**Comment by 24tngus** [comment]
확인했습니다!! (마지막 부분에 2번이랑 6번에 오타있써용!)

**Reply by jindream6128** [reply]
허거덩 확인했습니다 죄송함다 !!

**Comment by berryberries** [comment]
확인했습니다!

**Comment by jindream6128** [comment]
회사 내부의 네트워크 구성의경우, 네트워크 엔지니어의 영역인거같고 답변을 듣기 어려웠어용 ㅠㅠ
